tests/test_modeling_realm.py

```diff
@@ -0,0 +1,359 @@
+        model.to(torch_device)
+        model.eval()
+        result = model(input_ids, attention_mask=input_mask, token_type_ids=token_type_ids)
+        self.parent.assertEqual(result.projected_score.shape, (self.batch_size, self.retriever_proj_size))
+
+    def create_and_check_encoder(
+        self,
+        config,
+        input_ids,
+        token_type_ids,
+        input_mask,
+        candidate_inputs,
+        sequence_labels,
+        token_labels,
+        choice_labels,
+    ):
+        model = RealmEncoder(config=config)
+        model.to(torch_device)
+        model.eval()
+        relevance_score = floats_tensor([self.batch_size, self.num_candidates])
+        result = model(
+            candidate_inputs[0],
+            attention_mask=candidate_inputs[1],
+            token_type_ids=candidate_inputs[2],
+            relevance_score=relevance_score,
+            labels=token_labels,
+        )
+        self.parent.assertEqual(
+            result.logits.shape, (self.batch_size * self.num_candidates, self.seq_length, self.vocab_size)
+        )
+
+    def create_and_check_retriever(
+        self,
+        config,
+        input_ids,
+        token_type_ids,
+        input_mask,
+        candidate_inputs,
+        sequence_labels,
+        token_labels,
+        choice_labels,
+    ):
+        model = RealmRetriever(config=config)
+        model.to(torch_device)
+        model.eval()
+        result = model(
+            input_ids,
+            attention_mask=input_mask,
+            token_type_ids=token_type_ids,
+            candidate_input_ids=candidate_inputs[0],
+            candidate_attention_mask=candidate_inputs[1],
+            candidate_token_type_ids=candidate_inputs[2],
+        )
+        self.parent.assertEqual(result.relevance_score.shape, (self.batch_size, self.num_candidates))
+        self.parent.assertEqual(result.query_score.shape, (self.batch_size, self.retriever_proj_size))
+        self.parent.assertEqual(
+            result.candidate_score.shape, (self.batch_size, self.num_candidates, self.retriever_proj_size)
+        )
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            input_ids,
+            token_type_ids,
+            input_mask,
+            candidate_inputs,
+            sequence_labels,
+            token_labels,
+            choice_labels,
+        ) = config_and_inputs
+        inputs_dict = {"input_ids": input_ids, "token_type_ids": token_type_ids, "attention_mask": input_mask}
+        return config, inputs_dict
+
+
+@require_torch
+class RealmModelTest(ModelTesterMixin, unittest.TestCase):
+
+    all_model_classes = (
+        (
+            RealmEmbedder,
+            RealmEncoder,
+            # RealmRetriever is excluded from common tests as it is a container model
+            # consisting of two RealmEmbedders & simple inner product calculation.
+            # RealmRetriever
```

Comment:
Should we have a tester for it overwriting the common tests that are specific to this model?

Reply:
There are ten common tests that need to be overwritten, most of which failed because they were designed to contain one Transformer-like base model whereas `RealmRetriever` contains two. Additionally, the base models, `RealmEmbedder`, have been covered in the common tests. 

Therefore, I perceive overwriting these tests in order to fit `RealmRetriever` seems redundant. What do you think?

```
10 failed
         - tests/test_modeling_common.py:388 RealmModelTest.test_attention_outputs
         - tests/test_modeling_common.py:308 RealmModelTest.test_determinism
         - tests/test_modeling_common.py:1047 RealmModelTest.test_feed_forward_chunking
         - tests/test_modeling_common.py:711 RealmModelTest.test_headmasking
         - tests/test_modeling_common.py:943 RealmModelTest.test_hidden_states_output
         - tests/test_modeling_common.py:1315 RealmModelTest.test_inputs_embeds
         - tests/test_modeling_common.py:1173 RealmModelTest.test_model_common_attributes
         - tests/test_modeling_common.py:1247 RealmModelTest.test_model_outputs_equivalence
         - tests/test_modeling_common.py:1070 RealmModelTest.test_resize_tokens_embeddings
         - tests/test_modeling_common.py:142 RealmModelTest.test_save_load
```

Reply:
Fine with not overwriting them (we also don't use the common tests for RAG & EncoderDecoder)